#***********************************#
# ATAC-Seq pipeline - PAIR-END      #
# script: 10_consensus_peaks.Rmd    #
#***********************************

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Calling consensus peaks in Bioconductor
#https://ro-che.info/articles/2018-07-11-chip-seq-consensus
## First, load the Bioconductor libraries to work with genomic ranges:
```{r}
library(rtracklayer)
library(GenomicRanges)
```

##Get the list of .narrowPeak files in the current directory. We’ll treat them all as replicates. The same code should work with .bed or .gff files.
```{r}
peak_files_ProB_progenitors <- list.files(pattern = "*ProB-progenitors*")
peak_files_ProB_progenitors
```

## Read these files using rtracklayer’s import function:
```{r}
peak_granges_ProB_progenitors <- lapply(peak_files_ProB_progenitors, import)
peak_granges_ProB_progenitors
```

##peak_ranges is a simple R list with 3 elements, which individually are GRanges objects. Now let’s turn this list into a GRangesList, which in this case is more convenient to work with.
```{r}
peak_grangeslist_ProB_progenitors <- GRangesList(peak_granges_ProB_progenitors)
peak_grangeslist_ProB_progenitors
```

## Now, find the genome regions which are covered by at least 2 of the 3 sets of peaks:
```{r}
peak_coverage_ProB_progenitors <- coverage(peak_grangeslist_ProB_progenitors)
peak_coverage_ProB_progenitors
```

##peak_coverage is a list of numbers for every single genome position, encoded into an efficient structure Rle. Now, using the slice function, let’s get the regions in the genome where coverage is at least 2 or n-1:
```{r}
covered_ranges_ProB_progenitors <- slice(peak_coverage_ProB_progenitors, lower=2 , rangesOnly=T)
covered_ranges_ProB_progenitors
```

##This is a simple IRangesList object; let’s covert it to a GRanges object:
```{r}
covered_granges_ProB_progenitors <- GRanges(covered_ranges_ProB_progenitors)
covered_granges_ProB_progenitors
```

##And we are done! We can also save these regions as a .bed file using rtracklayer’s export:
```{r}
export(covered_granges_ProB_progenitors, "consensus_ProB_progenitors.bed")
```

Merging nearby peaks
##As I mentioned at the beginning, you might want to merge the peaks that are close to each other to avoid having many small fragments. In Bioconductor, you can do that using the min.gapwidth parameters to reduce. For example, this will merge the peaks that are separated by no more than 30 bp:
```{r}
covered_granges_ProB_progenitors_reduced <- reduce(covered_granges_ProB_progenitors, min.gapwidth=31)
export(covered_granges_ProB_progenitors_reduced, "consensusReduced_ProB_progenitors.bed")

```

When there's a peak in at least 2 samples 
CLLP:As you can see, this reduced the number of ranges from 63693  to 63429.
HSSC: As you can see, this reduced the number of ranges from 64704 to 64538
PrreB: As you can see, this reduced the number of ranges from 45953 to 45819 
PrroB: As you can see, this reduced the number of ranges from 53757 to 53533 
Immmature.B: As you can see, this reduced the number of ranges from 41926 to 41806
neuutrophils: As you can see, this reduced the number of ranges from 16776 to 16739
PrreProoB: As you can see, this reduced the number of ranges from 50700 to 50548


When there's a peak in at least n-1 samples (e.g 9 out of 10)
CLLP(8):As you can see, this reduced the number of ranges from 28095 to 28030.
HSSC(8): As you can see, this reduced the number of ranges from 31172 to 31130
PrreB(9): As you can see, this reduced the number of ranges from 23936 to 23886
PrroB(9): As you can see, this reduced the number of ranges from 28276 to 28204
Immmature.B(7): As you can see, this reduced the number of ranges from 23956 to 23907
neuutrophils(2): As you can see, this reduced the number of ranges from 16776 to 16739
#PrreProoBs(5): As you can see, this reduced the number of ranges from 32299 to 32252
PrreProoBs(2): As you can see, this reduced the number of ranges from 41290 to 41242
ProB-progenitors(2): As you can see, this reduced the number of ranges from 40621 to 40581


######################
# Get the whole consensus peaks for all cell types.
######################

#list of "consensus" bed files to get the whole consensus peaks
#setwd("/Users/mashael/Dropbox_kaust/Thesis_project/consensus/bed")
```{r}
file_names_bed <- grep("*consensusReduced*",dir(), value=TRUE)
```

#for single-cell analysis
#file_names_bed <- grep(".bed",dir(), value=TRUE)[c(5,6,11)]
```{r}
peak_granges <- lapply(as.list(file_names_bed), import, format = "BED")
peak_granges

all_peak_grangeslist <- GRangesList(peak_granges)
all_peak_grangeslist
```


#Includes a total of 6 GRanges objects, corresponding to the 8 different cell types 


#Now, get the coverage
```{r}
peak_coverage <- coverage(all_peak_grangeslist)
peak_coverage
```

#peak_coverage is a list of numbers for every single genome position, encoded into an efficient structure Rle. Now, using the slice function, let?s get the regions in the genome where coverage is at least in 1 sample:
```{r}
covered_ranges <- slice(peak_coverage, lower=1, rangesOnly=TRUE)
covered_ranges
```

#This is a simple IRangesList object; let?s covert it to a GRanges object:
```{r}
covered_granges <- GRanges(covered_ranges)
covered_granges
```

#cat("************************\n",summary(covered_granges),"\n************************\n")

#Merging nearby peaks
#you might want to merge the peaks that are close to each other to avoid having many small fragments. In Bioconductor, you can do that using the min.gapwidth parameters to reduce. For example, this will merge the peaks that are separated by no more than 30 bp
```{r}
red_covered_granges <- reduce(covered_granges, min.gapwidth=31)
red_covered_granges
```

#cat("************************\n",summary(red_covered_granges),"\n************************\n")

```{r}
export(red_covered_granges, "Allconsensus_cell_types.bed")

```

# We have a total of 61879 (50553_old) intervals (peaks)
# consensus_cell_types.bed = 53568 
# Allconsensus_cell_types.bed + PC = 53675 

#for singel-cell comparison
#export(red_covered_granges, "consensus_3cell_types.bed")